wallet/src/de/schildbach/wallet/ui/dashpay/ContactsActivity.kt

```diff
@@ -0,0 +1,256 @@
+        this.contactsAdapter.itemClickListener = this
+
+        initViewModel()
+
+        search.addTextChangedListener(this)
+
+
+
+        searchContacts()
+    }
+
+    private fun initViewModel() {
+        dashPayViewModel = ViewModelProvider(this).get(DashPayViewModel::class.java)
+        dashPayViewModel.searchContactsLiveData.observe(this, Observer {
+            if (Status.LOADING == it.status) {
+                startLoading()
+            } else {
+                stopLoading()
+                if (it.data != null) {
+                    processResults(it.data)
+                } else {
+                    showEmptyResult()
+                }
+            }
+        })
+        AppDatabase.getAppDatabase().blockchainIdentityDataDao().load().observe(this, Observer {
+            if (it != null) {
+                //TODO: we don't have an easy way of getting the identity id (userId)
+                val tx = walletApplication.wallet.getTransaction(it.creditFundingTxId)
+                val cftx = walletApplication.wallet.getCreditFundingTransaction(tx)
+                blockchainIdentityId = cftx.creditBurnIdentityIdentifier.toStringBase58()
+            }
+        })
+    }
+
+    private fun processResults(data: List<UsernameSearchResult>) {
+
+        val results = ArrayList<ContactSearchResultsAdapter.ViewItem>()
+        // process the requests
+        val requests = if (mode != MODE_SELECT_CONTACT)
+            data.filter { r -> r.isPendingRequest }.toMutableList()
+        else ArrayList()
+
+        val requestCount = requests.size
+        if (mode != MODE_VIEW_REQUESTS) {
+            while (requests.size > 3) {
+                requests.remove(requests[requests.size - 1])
+            }
+        }
+
+        if (requests.isNotEmpty() && mode != MODE_VIEW_REQUESTS)
+            results.add(ContactSearchResultsAdapter.ViewItem(null, ContactSearchResultsAdapter.CONTACT_REQUEST_HEADER, requestCount = requestCount))
+        requests.forEach { r -> results.add(ContactSearchResultsAdapter.ViewItem(r, ContactSearchResultsAdapter.CONTACT_REQUEST)) }
+
+        // process contacts
+        val contacts = if (mode != MODE_VIEW_REQUESTS)
+            data.filter { r -> r.requestSent && r.requestReceived }
+        else ArrayList()
+
+        if (contacts.isNotEmpty())
+            results.add(ContactSearchResultsAdapter.ViewItem(null, ContactSearchResultsAdapter.CONTACT_HEADER))
+        contacts.forEach { r -> results.add(ContactSearchResultsAdapter.ViewItem(r, ContactSearchResultsAdapter.CONTACT)) }
+
+        contactsAdapter.results = results
+        if (data.isEmpty()) {
+            showEmptyResult()
+        } else {
+            hideEmptyResult()
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        menuInflater.inflate(R.menu.contacts_menu, menu)
+        return super.onCreateOptionsMenu(menu)
+    }
+
+    private fun startLoading() {
+
```

Comment:
There is no loading screen in the designs, but I noticed a delay when the screen is first started up.

Since these methods are empty, I will remove them.

Reply:
https://github.com/dashevo/dash-wallet/pull/430/commits/121e84226bfb51599678b33b24338b88caaa9b97